auth/src/main/java/org/aerogear/auth/impl/AbstractAuthenticator.java

```diff
@@ -0,0 +1,32 @@
+package org.aerogear.auth.impl;
+
+import org.aerogear.auth.AuthenticationException;
+import org.aerogear.auth.credentials.ICredential;
+
+import java.security.Principal;
+
+/**
+ * Base class for all authenticators
+ */
+public class AbstractAuthenticator {
+
+    /**
+     * This method must be overridden with the custom authentication for the given credential.
+     *
+     * @param username username to be authenticated
+     * @param credential user credential
+     * @return the authenticated principal
+     * @throws AuthenticationException
+     */
+    public Principal authenticate(final String username, ICredential credential) throws AuthenticationException {
+        throw new IllegalStateException("Not implemented");
+    }
+
+    /**
+     * Logout the given principal
+     * @param principal principal to be log out
+     */
+    public void logout(Principal principal) {
```

Comment:
We will need to take in the identityToken as a param here. The logout URL can be constructed from some keycloak config constants, so we probably fetch them from within the method.

Reply:
We will pass here the Principal object so that we can have all the three tokens at once and any other information we want to put inside the Principal object

Reply:
Cool, that makes sense too.